Wikipedia/Networking/Fetchers/SavedArticlesFetcher.m

```diff
@@ -27,59 +24,188 @@ @implementation SavedArticlesFetcher
+                                                        .thenOn(self.accessQueue, ^(MWKArticle* article){
+                [[article allImageURLs] bk_each:^(NSURL* imageURL) {
+                    // fetch all article images, but don't block success on their completion or consider failed image
+                    // download an error
+                    [self.imageController downloadImageDataInBackground:imageURL];
+                }];
+                [self didFetchArticle:article title:title error:nil];
+            })
+                                                        .catch(^(NSError* error){
+                dispatch_async(self.accessQueue, ^{
+                    [self didFetchArticle:nil title:title error:error];
+                });
+            });
         }
     });
 }
 
 - (void)cancelFetch {
-    [self.fetcher cancelAllFetches];
+    [self.articleFetcher cancelAllFetches];
+}
+
+#pragma mark - KVO
+
+- (void)savedPageListDidChange:(NSDictionary*)change {
+    switch ([change[NSKeyValueChangeKindKey] integerValue]) {
+        case NSKeyValueChangeInsertion: {
+            [self savedPageListDidAddItems:change[NSKeyValueChangeNewKey]];
+            break;
+        }
+        case NSKeyValueChangeRemoval: {
+            [self savedPageListDidDeleteItems:change[NSKeyValueChangeOldKey]];
+            break;
+        }
+        case NSKeyValueChangeSetting: {
+            [self mergeFetchesWithUpdatedSavedPageList];
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+- (void)savedPageListDidDeleteItems:(NSArray<MWKSavedPageEntry*>*)deletedEntries {
+    if (!deletedEntries.count) {
+        return;
+    }
+    @weakify(self);
+    dispatch_async(self.accessQueue, ^{
+        @strongify(self);
+        BOOL wasFetching = self.fetchOperationsByArticleTitle.count > 0;
+        [deletedEntries bk_each:^(MWKSavedPageEntry* entry) {
+            DDLogInfo(@"Canceling saved page download for title: %@", entry.title);
+            [self.articleFetcher cancelFetchForPageTitle:entry.title];
+            [[[self.savedPageList.dataStore existingArticleWithTitle:entry.title] allImageURLs] bk_each:^(NSURL* imageURL) {
+                [self.imageController cancelFetchForURL:imageURL];
+            }];
+            [self.fetchOperationsByArticleTitle removeObjectForKey:entry.title];
+        }];
+        if (wasFetching) {
+            /*
+               only notify delegate if deletion occurs during a download session. if deletion occurs
+               after the fact, we don't need to inform delegate of completion
+             */
+            [self notifyDelegateIfFinished];
+        }
+    });
+}
+
+- (void)savedPageListDidAddItems:(NSArray<MWKSavedPageEntry*>*)insertedEntries {
+    if (!insertedEntries.count) {
+        return;
+    }
+    [self fetchTitles:[insertedEntries valueForKey:WMF_SAFE_KEYPATH([MWKSavedPageEntry new], title)]];
+}
+
+- (void)mergeFetchesWithUpdatedSavedPageList {
```

Comment:
Extra logic (and complexity) for dealing with a situation that in practice will never occur only because we exposed the savedPageList as readwrite in the public API.